### PR TITLE
Add the ability to disable/enable the context menu

### PR DESCRIPTION
**Pull request by loopmode**
This PR adds a dispose function and fixes https://github.com/sindresorhus/electron-context-menu/issues/82 and fixes https://github.com/sindresorhus/electron-context-menu/issues/84.

The implementation is such that the `contextMenu()` return value is a function that will remove the registered event listeners.

### CODE REVIEW

**Comment by loopmode** [comment]
@sindresorhus I also have updated the fixture.js and html files, but I'm not sure whether I did it "correctly", so it's parked on a separate branch. See https://github.com/loopmode/electron-context-menu/commit/26f36c4e6d9a2d709b57dfcbdee038d3fe190675

It adds a checkbox to enable or disable the context menu, but for that to work while keeping it simple, I moved the entire logic our of the js and into the html file.

I can provide another PR for that.

File: index.js [hunk 1]
```diff
@@ -244,29 +244,51 @@ const create = (win, options) => {
 			*/
 			menu.popup(electron.remote ? electron.remote.getCurrentWindow() : win);
 		}
-	});
+	};
+
+	webContents(win).on('context-menu', onContextMenu);
+	return () => webContents(win).removeListener('context-menu', onContextMenu);
```
Comment (sindresorhus) [inline index.js]: I prefer not using an inline arrow function when there's no return value. It just makes the code harder to read. Applies in other places too.

File: index.js [hunk 2]
```diff
@@ -244,29 +244,51 @@ const create = (win, options) => {
...
+			return;
+		}
+
+		const disposeMenu = create(win, options);
+		disposables.push(disposeMenu);
+	};
+
+	const dispose = () => {
+		disposables.forEach(dispose => dispose());
+		isDisposed = true;
```
Comment (sindresorhus) [inline index.js]: Instead of a boolean, just empty the array and check the length. You should empty the array regardless as you don't want to hold onto those forever.

**Comment by sindresorhus** [comment]
This looks great :)

You also need to add it to https://github.com/sindresorhus/electron-context-menu/blob/master/index.d.ts

**Comment by sindresorhus** [comment]
> It adds a checkbox to enable or disable the context menu, but for that to work while keeping it simple, I moved the entire logic our of the js and into the html file.

I think you should create a new fixture just for this. It's totally fine that it's in the HTML file. It can be as just `contextMenu()`, with no options. It should be part of this PR.

**Comment by loopmode** [comment]
Great, will do that!

**Comment by sindresorhus** [comment]
Bump :)

File: index.js (hunk 1 above)
Comment (loopmode) [reply index.js]: Returning a named `function dispose() { ...` is clearly the better option, also when debugging etc.

File: index.js (hunk 2 above)
Comment (loopmode) [reply index.js:269]: Hmm. I believe it would cause a chicken/egg problem and we'd never create the menu in the first place. Looking into it.

Comment (loopmode) [reply index.js:269]: Indeed, it's not that easy. 
Avoiding the flag requires either a completely different approach, or several more changes that seem unnecessarily complex.
It's really about bailing out in certain deferred edge cases, and the significant piece of information of the flag is: "No need to create the menu, the callsite already disposed it".

I already tried a couple variants, and I still believe the flag is the best solution. 

Should let go of the references despite of that.

**Comment by loopmode** [comment]
Hey @sindresorhus 

So.. I did quite some things extra here, hope it doesn't make accepting the PR too much of a problem?

- I added a .prettierrc which basically resembles the existing code style, at least I was able to format on save without causing extra changes
- I added a fixture for creating and disposing the menu, it can be used via `yarn start-fixture3`. I suggest you/we rename the fixture scripts :)
- Because the fixture files started to dominate the project root, i moved all of them into a `fixtures/` subfolder and adjusted the start-scripts accordingly
- The array of disposable functions does get emptied on dispose now, but I still did leave the `isDisposed` flag. As mentioned previously, it's currently essential. Maybe it's my lack of imagination, but I don't see another simple or elegant approach. Emptying the array and checking for length === 0 is unfortunately definitely not the way to go, as it cannot distinguish between "already disposed" and "not initialized yet". I'd be happy to learn something here, but really I think the boolean flag is the way to go

File: readme.md [hunk 3]
```diff
@@ -52,10 +52,21 @@ let mainWindow;
 })();
 ```
 
+Note that the return value of `contextMenu()` is a dispose function:
```
Comment (sindresorhus) [inline readme.md:55]: The readme and index.d.ts should be in sync documentation-wise.

File: index.js [hunk 4]
```diff
@@ -217,14 +217,12 @@ const create = (win, options) => {
...
-					visible: hasText && props.misspelledWord,
-					enabled: false
-				}
-			);
+			dictionarySuggestions.push({
+				id: 'dictionarySuggestions',
+				label: 'No Guesses Found',
+				visible: hasText && props.misspelledWord,
+				enabled: false
+			});
```
Comment (sindresorhus) [inline index.js:225]: Don't do unrelated changes.

**Comment by sindresorhus** [comment]
> I added a .prettierrc which basically resembles the existing code style, at least I was able to format on save without causing extra changes

I'm not interested in having this.

**Comment by sindresorhus** [comment]
> I added a fixture for creating and disposing the menu, it can be used via yarn start-fixture3. I suggest you/we rename the fixture scripts :)
Because the fixture files started to dominate the project root, i moved all of them into a fixtures/ subfolder and adjusted the package scripts accordingly
The array of disposable functions does get emptied on dispose now, but I still did leave the isDisposed flag. As mentioned previously, it's currently essential. Maybe it's my lack of imagination, but I don't see another simple or elegant approach. Emptying the array and checking for length === 0 is unfortunately definitely not the way to go, as it cannot distinguish between "already disposed" and "not initialized yet". I'd be happy to learn something here, but really I think the boolean flag is the way to go

👍

File: readme.md (hunk 3 above)
Comment (loopmode) [reply readme.md:55]: There are a couple of differences, as I didn't know about the sync aspect. Will adjust.
Actually, there's even a react hook example in index.d.ts that is not in readme. 
I will remove that altogether as it feels out of scope here.

File: index.js (hunk 4 above)
Comment (loopmode) [reply index.js:225]: Yeah, sorry. Related to prettierrc - will disable auto-formatting in my editor.

**Comment by loopmode** [comment]
> > I added a .prettierrc which basically resembles the existing code style, at least I was able to format on save without causing extra changes
> 
> I'm not interested in having this.

removed

File: index.js (hunk 2 above)
Comment (loopmode) [reply index.js:336]: array length reset

File: readme.md (hunk 3 above)
Comment (loopmode) [reply readme.md]: > The readme and index.d.ts should be in sync documentation-wise.

Synced

File: index.js (hunk 4 above)
Comment (loopmode) [reply index.js]: > Don't do unrelated changes.

Reverted

File: package.json [hunk 5]
```diff
@@ -10,10 +10,12 @@
 		"email": "sindresorhus@gmail.com",
 		"url": "https://sindresorhus.com"
 	},
+	"contributors": ["Jovica Aleksic <jovica.aleksic@loopmode.de>"],
```
Comment (sindresorhus) [inline package.json:13]: I don't use this field. It's not actually used anywhere and it's creates inequality with contributors that doesn't add themselves there. I prefer to keep it simple and just not use it. Hope you understand :)

**Comment by sindresorhus** [comment]
Thanks :)

File: package.json (hunk 5 above)
Comment (loopmode) [reply package.json]: Absolutely no problem! :)
